pytorch_lightning/callbacks/quantization.py

```diff
@@ -0,0 +1,201 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Quantization
+^^^^^^^^^^^^
+
+"""
+from typing import Any, Callable, Optional, Sequence, Union
+
+import torch
+from torch.quantization import QConfig
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+def wrap_qat_forward_context(
+    quant_cb,
+    model: pl.core.LightningModule,
+    func: Callable,
+    trigger_condition: Optional[Union[Callable, int]] = None
+) -> Callable:
+    """
+    This decorator is used as context manager...
+    """
+
+    # todo: consider using registering hook before/after forward
+    def wrapper(data) -> Any:
+        _is_func_true = isinstance(trigger_condition, Callable) and trigger_condition(model.trainer)
+        _is_count_true = isinstance(trigger_condition, int) and quant_cb._forward_calls < trigger_condition
+        _quant_run = trigger_condition is None or _is_func_true or _is_count_true
+        # apply custom trigger
+        if _quant_run:
+            quant_cb._forward_calls += 1
+            data = model.quant(data)
+        data = func(data)
+        # apply custom trigger
+        if _quant_run:
+            data = model.dequant(data)
+        return data
+
+    return wrapper
+
+
+def wrap_quantize_forward_context(model: pl.core.LightningModule, func: Callable, quant, dequant) -> Callable:
+    """
+    This decorator is used as context manager...
+    """
+
+    # todo: consider using registering hook before/after forward
+    def wrapper(data) -> Any:
+        data = model.quant(data)
+        data = func(data)
+        data = model.dequant(data)
+        return data
+
+    return wrapper
+
+
+def _recursive_hasattr(obj, attribs, state=True):
+    """recursive check if model has some layers denoted with ."""
+    if '.' in attribs:
+        attrib, attribs = attribs.split('.', 1)
+        if hasattr(obj, attrib):
+            return _recursive_hasattr(getattr(obj, attrib), attribs, state)
+        else:
+            return False
+    else:
+        return state & hasattr(obj, attribs)
+
+
+class QuantizationAwareTraining(Callback):
```

Comment:
Any particular reason it is not just called `Quantization`?

Reply:
there are a few other ways how to do quantization and this comes from official PT docs

Reply:
Yeah keep it as is! the name is important

Reply:
Okay, fine by me :]